# Hello!


This repo will be a collection of all things Bigham lab! It'll be continuously updated with tips and tricks that may be helpful starting out and/or to look back on to. Here's what's currently written:
* UsingHoffman: basics of accessing and using the lab's cluster memory and scheduler resources
* WhereAreFiles: working document of often needed or important files
* AccessingRStudioServer: how to use Rstudio server on Hoffman
* markdownTemplate: quick outline for writing markdown like for Github repositories 
* RmarkdownTemplate: quick outline for writing Rmarkdown
* scripts/fst_script.Rmd: manual Fst calculations
* DatabaseCheatSheet: some helpful websites
* SelectionScanResources: selection scan literature
* Geography of Genetic Variants Browser: https://popgen.uchicago.edu/ggv/?data=%221000genomes%22&chr=1&pos=222087833
* About detecting positive selection: https://mccoy-lab.github.io/hgv_modules/haplotype-based-signatures.html


* Plink tutorials:
    * This website has some Plink tutorials and videos Iâ€™ve found useful! I want to say they have some toy data you could practice with. I think they might be geared towards GWAS, but it may still be helpful background https://www.colorado.edu/ibg/isg-workshop/online-lectures-and-practicals/common-rare-variant-association The plink documentation also has a tutorial with example data I think: https://zzz.bwh.harvard.edu/plink/tutorial.shtml
 * An Owner's Guide to the Human Genome by Jonathan Pritchard : https://web.stanford.edu/group/pritchardlab/HGbook.html
* Classic papers! https://docs.google.com/spreadsheets/d/19tup7apWod0mlPpDWwehSD_TFE8MSbQ5IIg8_NB7Kx8/edit?gid=0#gid=0
